File: apps/blueprint-foundation/backend/app/schemas/summary.py
"""Summary module schemas for blueprint upload reports."""
from pydantic import BaseModel, computed_field
from typing import Optional, List
from datetime import datetime
from uuid import UUID
from enum import Enum


class QualityGrade(str, Enum):
    """Overall quality assessment grade."""
    EXCELLENT = "EXCELLENT"
    GOOD = "GOOD"
    FAIR = "FAIR"
    POOR = "POOR"


class PageSummary(BaseModel):
    """Summary data for a single page."""
    page_id: UUID
    page_number: int
    dimensions: str
    dpi: Optional[int]
    is_calibrated: bool
    scale_info: Optional[str]
    warning_count: int
    warnings: List[str]

    class Config:
        from_attributes = True


class UploadSummary(BaseModel):
    """Comprehensive summary for a single upload."""
    upload_id: UUID
    project_id: UUID
    filename: str
    file_type: str
    file_size_mb: float
    status: str
    uploaded_at: datetime

    # Page metrics
    total_pages: int
    calibrated_pages: int
    pages_with_warnings: int

    # Quality assessment
    quality_grade: QualityGrade
    quality_score: int  # 0-100

    # Warnings aggregate
    total_warnings: int
    warning_types: dict[str, int]

    # Pages detail
    pages: List[PageSummary]

    # Processing info
    processing_duration_seconds: Optional[float]
    error_message: Optional[str]

    class Config:
        from_attributes = True


class ProjectSummary(BaseModel):
    """Aggregate summary for an entire project."""
    project_id: UUID
    project_name: str
    address: Optional[str]
    builder: Optional[str]
    foundation_type: str
    floors: int
    created_at: datetime

    # Upload metrics
    total_uploads: int
    completed_uploads: int
    failed_uploads: int
    processing_uploads: int

    # Page metrics
    total_pages: int
    calibrated_pages: int

    # Quality metrics
    average_quality_score: float
    uploads_with_warnings: int
    total_warnings: int

    # Storage
    total_size_mb: float

    # Uploads detail
    uploads: List[UploadSummary]

    class Config:
        from_attributes = True


class DashboardSummary(BaseModel):
    """High-level dashboard statistics."""
    total_projects: int
    total_uploads: int
    total_pages: int
    total_calibrations: int

    # Status breakdown
    uploads_by_status: dict[str, int]

    # Quality overview
    average_quality_score: float
    uploads_with_warnings: int

    # Storage
    total_storage_mb: float

    # Recent activity
    recent_uploads: List[UploadSummary]

    # Timestamp
    generated_at: datetime

    class Config:
        from_attributes = True


class ExportFormat(str, Enum):
    """Supported export formats."""
    JSON = "json"
    CSV = "csv"


class ExportRequest(BaseModel):
    """Request for exporting summary data."""
    format: ExportFormat = ExportFormat.JSON
    include_pages: bool = True
    include_warnings: bool = True

File: apps/blueprint-foundation/frontend/src/types/summary.ts
/**
 * Summary module type definitions
 */

export enum QualityGrade {
  EXCELLENT = 'EXCELLENT',
  GOOD = 'GOOD',
  FAIR = 'FAIR',
  POOR = 'POOR',
}

export interface PageSummary {
  page_id: string
  page_number: number
  dimensions: string
  dpi: number | null
  is_calibrated: boolean
  scale_info: string | null
  warning_count: number
  warnings: string[]
}

export interface UploadSummary {
  upload_id: string
  project_id: string
  filename: string
  file_type: string
  file_size_mb: number
  status: string
  uploaded_at: string
  total_pages: number
  calibrated_pages: number
  pages_with_warnings: number
  quality_grade: QualityGrade
  quality_score: number
  total_warnings: number
  warning_types: Record<string, number>
  pages: PageSummary[]
  processing_duration_seconds: number | null
  error_message: string | null
}

export interface ProjectSummary {
  project_id: string
  project_name: string
  address: string | null
  builder: string | null
  foundation_type: string
  floors: number
  created_at: string
  total_uploads: number
  completed_uploads: number
  failed_uploads: number
  processing_uploads: number
  total_pages: number
  calibrated_pages: number
  average_quality_score: number
  uploads_with_warnings: number
  total_warnings: number
  total_size_mb: number
  uploads: UploadSummary[]
}

export interface DashboardSummary {
  total_projects: number
  total_uploads: number
  total_pages: number
  total_calibrations: number
  uploads_by_status: Record<string, number>
  average_quality_score: number
  uploads_with_warnings: number
  total_storage_mb: number
  recent_uploads: UploadSummary[]
  generated_at: string
}

export type ExportFormat = 'json' | 'csv'

export interface ExportRequest {
  format: ExportFormat
  include_pages?: boolean
  include_warnings?: boolean
}
